### ISSUE TITLE
Consolidate .open() and .openAndYieldController()

### ISSUE THREAD

**Issue by StrahilKazlachev** [aurelia/dialog#224]
**I'm submitting a feature request**
- **Library Version:**
  1.0.0-beta.3.0.1

**Please tell us about your environment:**
- **Operating System:**
  Windows [7|10]
- **Node Version:**
  6.2.0
- **NPM Version:**
  3.8.9
- **JSPM OR Webpack AND Version**
  JSPM 0.16.32
- **Browser:**
  all
- **Language:**
  all

**Current behavior:**
Two methods to open a dialog:
- `.open(settings: DialogSettings): Promise<DialogResult>`
- `.openAndYieldController(setings: DialogSettings): Promise<DialogController & { result: Promise<DialogResult> }>`

**Expected/desired behavior:**
- **What is the expected behavior?**
  To consolidate the two methods as follows:
  - **rename** `DialogResult` => `CloseDialogResult`
  - add **new** dialog result interface

```
interface OpenDialogResult { 
  controller: DialogController | null;
  wasCancelled: boolean;
  result: Promise<CloseDialogResult>
}
```
- the merged `.open()` definition 

```
open(settings: DialogSettings): 
        { openResult: Promise<OpenDialogResult>, closeResult: Promise<CloseDialogResult> };
// OR
open(settings: DialogSettings & { yieldController: false | undefined }): Promise<CloseDialogResult>;
open(settings: DialogSettings & { yieldController: true }): Promise<OpenDialogResult>;
open(settings: DialogSettings & { yieldController: boolean | undefined }): Promise<OpenDialogResult | CloseDialogResult>;
```
- **What is the motivation / use case for changing the behavior?**
  First of all, I get that `openAndYieldController` was introduced so that new functionality could be added in non-breaking way, but
  - confusion when to use `.open()` and when `.openAndYieldController()` - I think there was at least 1 issue, also on the gitter channel, and at work(I shouldn't have said that).
  - adding the `result: Promise<DialogResult>` property on `DialogController` doesn't feel right, also it's somewhat hard to type currently #216
  - with the current impl `.open()` is a subset of `.openAndYieldController()`
  - in the context of #223 we can propagate cancellation in consistent way - having `.wasCancelled` in both `DialogOpenResult` and `DialogCloseResult`(cancelable promises do seem far 😞)

I'll make a PR to comment on actual code if these changes will be considered.

**Comment by EisenbergEffect** [comment]
I would be willing to consider them. I thin we would likely want to make some changes but if you are willing to put a PR together, we can use that as a point of conversation and refine from there.

**Comment by StrahilKazlachev** [comment]
Will do tomorrow, it's too late for me now.

**Comment by PWKad** [comment]
For reference - https://github.com/aurelia/dialog/pull/218

**Comment by PWKad** [comment]
@StrahilKazlachev @EisenbergEffect We've discussed going back to a promise that gets resolved or if not it would get thrown and need to be caught by the consuming developer.  Now that we are talking about working on open and what gets returned do we want to re-hash this conversation?

**Comment by StrahilKazlachev** [comment]
I went over #22, but I like the current approach, rejecting only under error conditions. At a point I experimented with an API that cancelled with rejection. Found it easier to reason cancellation on resolve rather then on reject.
We could add to the returned promises a `.throwOnError()` method to opt-in to throw on cancel:
```
service.open(settings)
  .then(() => { /* success or cancellation */ })
  .catch(() => { /* error */ })

service.open(settings).throwOnError()
  .then(() => { /* success */ })
  .catch(() => { /* cancellation or error */ })
``` 
But I don't like how it has to be typed, without union types.
Cancelable Promises - the proposal TC39 discusses is at stage 1 - I think, and there are a couple other proposals. I wouldn't go with them in the dialog now.

**Comment by StrahilKazlachev** [comment]
What do you think about a `throwOnCancel` flag in `DialogSettings`? But with a restriction:
* `false` by default
* the default can not be configured

This way if a plugin depends on the dialog it will not break.

**Comment by PWKad** [comment]
I personally like that approach, @EisenbergEffect seem reasonable?

**Comment by EisenbergEffect** [comment]
That sounds good to me.

**Comment by StrahilKazlachev** [comment]
Closed #225.
